# import statements

In [80]:
import pandas as pd
import numpy as np

In [4]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [5]:
df = pd.read_csv("C:/Users/ankit/Downloads/auto.csv", names = headers)

In [45]:
shp = df.shape
print(f"the table has {shp[0]} rows and {shp[1]} number of columns.")
df.head()

the table has 205 rows and 26 number of columns.


symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

# Dealing with missing values in the data set

In [8]:
df.replace("?", np.nan, inplace=True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

# evaluation of missing data is there any missing data??

In [20]:
missing_data = df.isnull()
missing_data.head()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

 # True indiacates the missing data in the particular columns

In [22]:
for column in missing_data.columns.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print(" ")

symboling
False    205
Name: symboling, dtype: int64
 
normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64
 
make
False    205
Name: make, dtype: int64
 
fuel-type
False    205
Name: fuel-type, dtype: int64
 
aspiration
False    205
Name: aspiration, dtype: int64
 
num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64
 
body-style
False    205
Name: body-style, dtype: int64
 
drive-wheels
False    205
Name: drive-wheels, dtype: int64
 
engine-location
False    205
Name: engine-location, dtype: int64
 
wheel-base
False    205
Name: wheel-base, dtype: int64
 
length
False    205
Name: length, dtype: int64
 
width
False    205
Name: width, dtype: int64
 
height
False    205
Name: height, dtype: int64
 
curb-weight
False    205
Name: curb-weight, dtype: int64
 
engine-type
False    205
Name: engine-type, dtype: int64
 
num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64
 
engine-size
False    205
Name: engine-size, dtype: int64

Based on the summary above, each column has 205 rows of data and seven of the columns containing missing data:
<ol>
    <li>"normalized-losses": 41 missing data</li>
    <li>"num-of-doors": 2 missing data</li>
    <li>"bore": 4 missing data</li>
    <li>"stroke" : 4 missing data</li>
    <li>"horsepower": 2 missing data</li>
    <li>"peak-rpm": 2 missing data</li>
    <li>"price": 4 missing data</li>
</ol>

# dealing with the missing data

In [23]:
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

In [27]:
NL_mean_value = df["normalized-losses"].astype("float").mean()
print(f"the mean normalized loss is: {mean_value}")

the mean normalized loss is: 122.0


In [28]:
df["normalized-losses"].replace(np.nan, NL_mean_value, inplace=True)

In [32]:
df.isnull()["normalized-losses"].value_counts()

False    205
Name: normalized-losses, dtype: int64

In [36]:
B_mean_v = df["bore"].astype("float").mean()
df["bore"].replace(np.nan, B_mean_v, inplace=True)
df.isnull()["bore"].value_counts()

False    205
Name: bore, dtype: int64

In [40]:
S_mean_v = df["stroke"].astype("float").mean()
df["stroke"].replace(np.nan, S_mean_v, inplace=True)
df.isnull().stroke.value_counts()

False    205
Name: stroke, dtype: int64

In [42]:
H_mean_v= df["horsepower"].astype("float").mean()
df["horsepower"].replace(np.nan, H_mean_v, inplace=True)
df.isnull().horsepower.value_counts()

False    205
Name: horsepower, dtype: int64

In [50]:
P_mean_v = df["peak-rpm"].astype("float").mean()
df["peak-rpm"].replace(np.nan, P_mean_v, inplace=True)
df.isnull()["peak-rpm"].value_counts()

False    205
Name: peak-rpm, dtype: int64

In [52]:
CV_frequency = df["num-of-doors"].value_counts()
CV_frequency

four    114
two      89
Name: num-of-doors, dtype: int64

In [53]:
df['num-of-doors'].replace(np.nan, "four", inplace=True)

In [54]:
df.isnull()["num-of-doors"].value_counts()

False    205
Name: num-of-doors, dtype: int64

In [55]:
df.dropna(subset=["price"], axis=0, inplace=True)
df.reset_index(drop=True, inplace=True)

In [66]:
missing_data = df.isnull()
for column in missing_data.columns.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print(" ")

symboling
False    201
Name: symboling, dtype: int64
 
normalized-losses
False    201
Name: normalized-losses, dtype: int64
 
make
False    201
Name: make, dtype: int64
 
fuel-type
False    201
Name: fuel-type, dtype: int64
 
aspiration
False    201
Name: aspiration, dtype: int64
 
num-of-doors
False    201
Name: num-of-doors, dtype: int64
 
body-style
False    201
Name: body-style, dtype: int64
 
drive-wheels
False    201
Name: drive-wheels, dtype: int64
 
engine-location
False    201
Name: engine-location, dtype: int64
 
wheel-base
False    201
Name: wheel-base, dtype: int64
 
length
False    201
Name: length, dtype: int64
 
width
False    201
Name: width, dtype: int64
 
height
False    201
Name: height, dtype: int64
 
curb-weight
False    201
Name: curb-weight, dtype: int64
 
engine-type
False    201
Name: engine-type, dtype: int64
 
num-of-cylinders
False    201
Name: num-of-cylinders, dtype: int64
 
engine-size
False    201
Name: engine-size, dtype: int64
 
fuel-system
False    20

# No any missing data no any uncleaned and un matched data just pure data with clean values.... now its time for changing the data types of the column and normalization

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

# we can see the unmatched data types for the columns we will try to mend it..

In [69]:
df[["bore", "stroke"]] = df[["bore", "stroke"]].astype("float")

In [70]:
df[["normalized-losses"]] = df[["normalized-losses"]].astype("int")

In [71]:
df[["price"]] = df[["price"]].astype("float")

In [72]:
df[["peak-rpm"]] = df[["peak-rpm"]].astype("float")

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    int32  
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       201 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

# it seems good..

# Now for the normalization part

# it is just a process of changing the range of column value from 0 to 1 which indicates the total proportion of how much dies the each value contributes to the each row

In [75]:
df.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')

In [76]:
maximum_len = df['length'].max()
maximum_wid = df['width'].max()
maximum_hei = df['height'].max()

In [77]:
df["length"] = df["length"]/maximum_len
df["width"] = df["width"]/maximum_wid
df["height"] = df["height"]/maximum_hei

In [79]:
df[['length', 'width', 'height']].head()

length     width    height
0  0.811148  0.890278  0.816054
1  0.811148  0.890278  0.816054
2  0.822681  0.909722  0.876254
3  0.848630  0.919444  0.908027
4  0.848630  0.922222  0.908027

# Normalized

# NOW OUR DATA IS CLEANED, NORMALIZED AND SET FOR EXPLORITORY DATA ANALYSIS.(DATA PRE-PROCESSING PART)